ADR/0033-upstreaming-integration-service.md

```diff
@@ -0,0 +1,270 @@
+        version: tekton.dev/v1
+    labels:
+      match:
+        "appstudio.openshift.io/application": my-application
+        "pipelinesascode.tekton.dev/state": completed
+  extractor:
+    image_url: '.status.results.[] | select(.name == "IMAGE_URL").value'
+    image_digest: '.status.results.[] | select(.name == "IMAGE_DIGEST").value'
+    name: '.metadata.labels."appstudio.openshift.io/component"'
+    source:
+      git:
+        revision: '.metadata.labels."pipelinesascode.tekton.dev/sha"'
+        url: '.metadata.labels."pipelinesascode.tekton.dev/repo-url"'
+```
+
+Don't take the fields literally. The details will likely change after the ADR has settled. It is
+here for illustration purposes.
+
+The `selector` field will be used to identify resources that should trigger the construction of new
+[TestSubjects]. When created, the *basis* for the new test subject should always be taken from the
+**control** [TestSubject]. The values extracted from the triggering resource by way of the
+expressions on the `extractor` resource should be applied to that basis in the construction of a new
+[TestSubject]. If there is no **control** [TestSubject], then the new TestSubject should be created
+containing only one element: the new pullspec.
+
+In the above example, the constructor will select only pipelineruns which are completed and which
+are associated with the app `my-application`. When triggered, it will extract the same image and
+source details that we extract today to supply to a new [TestSubject].
+
+### Release Initiation, as owned by the release-service
+
+For the purpose of this ADR, we will only say that release-service will take over the
+responsibility of initiating Releases. That will be described in more detail in a subsequent,
+forthcoming ADR.
+
+Note, in this pattern, [Release] construction looks a lot like [TestSubject] construction. The
+triggered system knows enough about the triggering system to optionally initiate itself.
+
+## Use Cases in Detail
+
+**Using integration-service as a remote API from Jenkins.** A hypothetical platform team might want
+to make use of [integration-service] from a jenkins pipeline. For example, perhaps their jenkins job
+performs CI on a git repository that contains pullspecs which should be mirrored to some environment
+of theirs. Before mirroring, they want to make sure that the same test works against dozens of
+different kubernetes configurations on different public clouds. That platform team first
+creates a single tekton pipeline that contains the logic to perform the test given a single cluster.
+They then configure dozens of [IntegrationTestScenario] resources, one for each kubernetes
+configuration they want to target. When a new set of pullspecs is proposed to their git repo,
+Jenkins responds. Jenkins then creates a single [TestSubject] in the kubernetes namespace,
+representing the new set of images under test. [integration-service] responds, inspecting all of the
```

Comment:
This is counter to the proposed TestSubject creation above, namely that the integration service is no longer responsible for creating the TestSubjects. How can the integration service validate that a TestSubject is a legitimate request? Should the integration service care about what is _inside_ the TestSubject or should it just accept _anything_ and forward it on to users' provided ITSs?

Reply:
Oh yeah - that's just the point though, that automatic TestSubject creation is an optional behavior of itnegration-service which can be configured, turned off or on by the absence or presence of a TestSubjectConstructor.

What would an illegitimate TestSubject look like?

Reply:
Remember, not just anybody can create TestSubjects. They have to have RBAC in the namespace where the ITS resides that permits them to create TestSubjects. In this jenkins case, imagine that jenkins has been provisioned a serviceaccount that has explicit rights to create TestSubjects.

Of course, if that SA token were stolen, an attack could create a TestSubject with references to an image of theirs that runs a bitcoin miner (or whatever).

Reply:
> What would an illegitimate TestSubject look like?

I think the situation that I was considering is not likely to be realized. If TestSubjects are manually created _and_ the same TestSubjects (i.e. with the same `LABELNAME`) are created with a TestSubjectConstructor, the two might not have the same data. The manually created ones might not be consistent with extracted data. Does this matter? I assume that the answer is "no" as it would be up to the user/process creating the TestSubjects that it is properly formed for the ITS configured to run from it.

Reply:
I agree that manually created ones shouldn't necessarily need to correspond to generated ones, this kind of behavior is supported by the integration service even now where manually created Snapshots are still tested and processed.